libs-scala/gatling-utils/src/main/scala/com/daml/gatling/stats/SimulationLog.scala

```diff
@@ -0,0 +1,273 @@
+              "count" -> stats.count.toString,
+              "successCount" -> stats.successful.count.toString,
+              "errorCount" -> stats.failed.count.toString,
+              "min" -> format(stats.successful.percentile(0.0)),
+              "p90" -> format(stats.successful.percentile(0.9)),
+              "p95" -> format(stats.successful.percentile(0.95)),
+              "p99" -> format(stats.successful.percentile(0.99)),
+              "p999" -> format(stats.successful.percentile(0.999)),
+              "max" -> format(stats.successful.percentile(1.0)),
+              "mean" -> format(stats.successful.geometricMean.map(math.round)),
+              "avg" -> format(stats.successful.mean.map(math.round)),
+              "stddev" -> format(stats.successful.stdDev.map(math.round)),
+              "rps" -> format(stats.successful.requestsPerSecond)
+            )
+        }
+      }
+  }
+}
+
+object SimulationLog {
+  type Timestamp = Long
+
+  private def format[A: Numeric: Show](fa: Option[A]): String = {
+    import scalaz.syntax.show._
+    val num = implicitly[Numeric[A]]
+    val zero: A = num.fromInt(0)
+    fa.getOrElse(zero).shows
+  }
+
+  case class ScenarioStats(
+      label: String,
+      maxUsers: Int = 0,
+      requestsByType: Map[String, RequestTypeStats] = Map.empty)
+
+  case class DurationStatistics(
+      durations: Seq[Int],
+      start: Option[Timestamp],
+      end: Option[Timestamp]) {
+    def count: Int = durations.size
+    def mean: Option[Double] = durations.nonEmptyOpt.map(ds => ds.sum.toDouble / ds.size)
+    def geometricMean: Option[Double] =
+      durations.nonEmptyOpt.map(ds => math.exp(ds.map(d => math.log(d.toDouble)).sum / ds.size))
+    def duration: Option[Int] = for { s <- start; e <- end } yield (e - s).toInt
+    def requestsPerSecond: Option[Double] = duration.map(count.toDouble / _.toDouble * 1000)
+
+    def stdDev: Option[Double] =
+      for {
+        avg <- mean
+        variance <- durations.nonEmptyOpt.map(ds => ds.map(d => math.pow(d - avg, 2)).sum / ds.size)
+      } yield math.sqrt(variance)
+
+    def percentile(p: Double): Option[Int] = {
+      require(p >= 0.0 && p <= 1.0, "Percentile must be between zero and one, inclusive.")
+      sortedDurations.nonEmptyOpt.map(ds => ds(Math.round((ds.size - 1).toDouble * p).toInt))
+    }
+
+    private lazy val sortedDurations = durations.toIndexedSeq.sorted
+  }
+
+  object DurationStatistics {
+    implicit val durationStatisticsMonoid: Monoid[DurationStatistics] =
+      new Monoid[DurationStatistics] {
+        override def zero: DurationStatistics = DurationStatistics(Seq.empty, None, None)
+
+        override def append(s1: DurationStatistics, s2: => DurationStatistics): DurationStatistics =
+          DurationStatistics(
+            s1.durations ++ s2.durations,
+            Tag.unwrap(Tags.Min(s1.start) |+| Tags.Min(s2.start)),
+            Tag.unwrap(Tags.Max(s1.end) |+| Tags.Max(s2.end))
```

Comment:
Your instance is unlawful as written, because in the min semigroup, `None < Some(a)` for all _a_, but you have declared `None` to be your monoid's zero. Instead, you want the Min applied directly to Duration, then lifted into the option monoid:

```suggestion
            Tag.unsubst(Tags.Min.subst(s1.start) |+| Tags.Min.subst(s2.start)),
            Tag.unsubst(Tags.Max.subst(s1.end) |+| Tags.Max.subst(s2.end))
```

In other words you want to append two `Option[Duration @@ Tags.Min]`, _not_ two `Option[Duration] @@ Tags.Min`.

You may find it more expedient to simply leave the `@@ Tags.Min` and max in the correct place in the declaration of `DurationStatistics`; do whichever is more convenient.

Reply:
ummmm... why not
```
            s1.start |+| s2.start,
            s1.end |+| s2.end
```
I get the point, that we want to apend Min and Min, Max and Max... but in this particular case it does not make anything safer. We are wrapping, appending and unwrapping.... So if you wrap the wrong thing you get the wrong thing. Just need to be careful what appends with what.